src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1138 @@
+            {
+               Vst::BusInfo busInfo;
+               if(effect->getBusInfo(Vst::kAudio, Vst::kOutput, busIndex, busInfo) != kResultOk)
+               {
+                  return 0;
+               }
+               if(busInfo.busType == Vst::kMain)
+               {
+                  data.outputs[busIndex].numChannels = busInfo.channelCount;
+                  data.outputs[busIndex].channelBuffers32 = outBlock + outputBlocksOffset;
+                  outputBlocksOffset += busInfo.channelCount;
+               }
+               else
+               {
+                  //aux is not yet supported
+                  data.outputs[busIndex].numChannels = 0;
+                  data.outputs[busIndex].channelBuffers32 = nullptr;
+               }
+               data.outputs[busIndex].silenceFlags = 0UL;
+            }
+         }
+      
+         const auto processResult = audioProcessor->process(data);
+      
+         return processResult == kResultOk ?
+            data.numSamples : 0;
+      }
+      return 0;
+   }
+}
+
+size_t VST3Effect::ProcessBlock(float** inBlock, float** outBlock, size_t blockLen)
+{
+   internal::ComponentHandler::PendingChangesPtr pendingChanges { nullptr };
+   if(mComponentHandler)
+      pendingChanges = mComponentHandler->getPendingChanges();
+   return VST3ProcessBlock(mEffectComponent.get(), mSetup, inBlock, outBlock, blockLen, pendingChanges.get());
+}
+
+bool VST3Effect::RealtimeInitialize()
+{
+   //reload current parameters form the editor into parameter queues
+   SyncParameters();
+   return true;
+}
+
+bool VST3Effect::RealtimeAddProcessor(unsigned numChannels, float sampleRate)
+{
+   using namespace Steinberg;
+
+   try
+   {
+      auto effect = std::make_unique<VST3Effect>(*this);
+      effect->mSetup.processMode = Vst::kRealtime;
+      effect->mSetup.sampleRate = sampleRate;
+      if(!effect->ProcessInitialize({0}, nullptr))
+         throw std::runtime_error { "VST3 realtime initialization failed" };
+
+      mRealtimeGroupProcessors.push_back(std::move(effect));
+      return true;
+   }
+   catch(std::exception& e)
+   {
+      //make_unique<> isn't likely to fail here since this effect instance
+      //somehow succeeded
+      wxLogError("Failed to add realtime processor: %s", e.what());
+   }
+   return false;
+}
+
+bool VST3Effect::RealtimeFinalize()
+{
+   for(auto& processor : mRealtimeGroupProcessors)
+      processor->ProcessFinalize();
+
+   mRealtimeGroupProcessors.clear();
+   mPendingChanges.reset();
+   
+   return true;
+}
+
+bool VST3Effect::RealtimeSuspend()
+{
+   for(auto& effect : mRealtimeGroupProcessors)
+      effect->mAudioProcessor->setProcessing(false);
+   return true;
+}
+
+bool VST3Effect::RealtimeResume()
+{
+   for(auto& effect : mRealtimeGroupProcessors)
+      effect->mAudioProcessor->setProcessing(true);
+   return true;
+}
+
+bool VST3Effect::RealtimeProcessStart()
+{
+   wxASSERT(mPendingChanges == nullptr);
```

Comment:
This function is called in a worker thread.  It's not good to call wxASSERT there.  If the assertion is violated for any reason, the debug build will try to make a dialog from a non-main thread.

Maybe just a simple `assert` instead if you must express some assertion.

Reply:
Yep, thanks